File: baekjoon/10703-유성/01.py
"""
문제:
    - 유성이 떨어지기 전 사진이 주어졌을 때, 유성이 떨어지고 난 후의 상황을 출력하라.
조건:
    - 사진은 유성, 땅, 공기로만 이루어져 있음.
    - 유성, 땅은 각각 하나의 덩어리임.
    - 떨어지고 난 후의 첫줄은 공기, 마지막줄은 땅으로 이루어져있음.
    - 모든 유성 조각들은 연결되어 있다. 즉, 두 부분 유성이 존재한다면, 한 쪽에서 유성
      조각을 통해 상하좌우로 이동해서 다른 부분 유성에 도달할 수 있다. 땅 또한 같은
      방식으로 연결되어 있다. -> 이거 뭔소린지 모르겠음
    - 유성은 수직으로 낙하한다. 유성이 땅에 떨어졌을 때, 유성과 땅은 원형을 유지한다
풀이:
    - 유성이 땅에 닿았느냐를 어떻게 판단할까?
        - 중간에 공기층이 없어야함
"""
from typing import List, Self
from sys import stdin, stdout
from enum import Enum

class PixelType(Enum):
    AIR = '.'
    SHOOTING_STAR = 'X'
    TERRAN = '#'

class Photo():
    terran_data: List[List[PixelType]]
    star_data: List[List[PixelType]]
    width: int
    height: int

    @classmethod
    def from_string(cls, data: str) -> Self:
        photo = Photo()
        data = [list(line) for line in data.split('\n')]
        width = len(data[0])
        height = len(data)
        terran_data = []
        star_data = []
        for i in range(height):
            terran_data.append([])
            star_data.append([])
            for j in range(width):
                if data[i][j] == PixelType.SHOOTING_STAR.value:
                    star_data[i].append(PixelType.SHOOTING_STAR)
                    terran_data[i].append(PixelType.AIR)
                elif data[i][j] == PixelType.TERRAN.value:
                    star_data[i].append(PixelType.AIR)
                    terran_data[i].append(PixelType.TERRAN)
                else:
                    star_data[i].append(PixelType.AIR)
                    terran_data[i].append(PixelType.AIR)
        cls.terran_data = terran_data
        cls.star_data = star_data
        cls.width = width
        cls.height = height
        return photo

    @classmethod
    def from_before(cls, before_photo: Self) -> Self:
        # find_minimum_distance
        distance = before_photo.get_star_distance()
        new_photo = cls()
        new_photo.terran_data = before_photo.terran_data
        new_photo.star_data = [
            [PixelType.AIR for _ in range(before_photo.width)] \
            for _ in range(before_photo.height)
        ]
        for i in range(new_photo.height):
            for j in range(new_photo.width):
                if before_photo.star_data[i][j] == PixelType.SHOOTING_STAR:
                    new_photo.star_data[i + distance][j] = PixelType.SHOOTING_STAR
        new_photo.width = before_photo.width
        new_photo.height = before_photo.height
        return new_photo

    def get_star_distance(self) -> int:
        distance = self.height
        for j in range(self.width):
            last_star_index = -1
            for i in range(self.height):
                if self.star_data[i][j] == PixelType.SHOOTING_STAR:
                    last_star_index = i
                elif self.terran_data[i][j] == PixelType.TERRAN:
                    if last_star_index >= 0:
                        distance = min(i - last_star_index - 1, distance)
                    break

        return distance

    def __str__(self) -> str:
        s = ''
        for i in range(self.height):
            for j in range(self.width):
                if self.star_data[i][j] == PixelType.SHOOTING_STAR:
                    s += self.star_data[i][j].value
                else:
                    s += self.terran_data[i][j].value
            s += '\n'
        return s

if __name__ == '__main__':
    stdin.readline()
    before_photo = Photo.from_string(stdin.read())
    after_photo = Photo.from_before(before_photo)
    stdout.write(str(after_photo))

File: baekjoon/10703-유성/01.rs
use std::io;

fn main() {
    let (height, width) = get_two_numbers();
    let before_photo = create_photo_from_stdin();
    let min_distance = before_photo.get_min_distance();

    let mut buff = String::new();
    for i in 0..height {
        for j in 0..width {
            if (i as isize) - min_distance >= 0 && before_photo.data[(i - min_distance) as usize][j as usize] == Pixel::Star {
                buff.push('X');
            } else if before_photo.data[i as usize][j as usize] == Pixel::Terran {
                buff.push('#');
            } else {
                buff.push('.');
            }
        }
        buff.push('\n');
    }

    println!("{}", buff);
}

fn get_two_numbers() -> (isize, isize) {
    let mut line = String::new();
    io::stdin().read_line(&mut line).unwrap();
    let mut iter = line.split_whitespace();
    let a = iter.next().unwrap().parse::<isize>().unwrap();
    let b = iter.next().unwrap().parse::<isize>().unwrap();
    (a, b)
}

#[derive(Copy, Clone, Debug, PartialEq)]
enum Pixel {
    Air,
    Terran,
    Star,
}

struct Photo {
    data: Vec<Vec<Pixel>>,
}

impl Photo {
    fn get_min_distance(&self) -> isize {
        let mut min_distance = isize::MAX;
        for j in 0..self.data[0].len() {
            let mut count = 0;
            let mut is_meet_star = false;
            for i in 0..self.data.len() {
                match self.data[i][j] {
                    Pixel::Air => {
                        count += 1;
                    }
                    Pixel::Terran => {
                        if is_meet_star {
                            min_distance = std::cmp::min(min_distance, count);
                        }
                        break;
                    }
                    Pixel::Star => {
                        count = 0;
                        is_meet_star = true;
                    }
                }
            }
        }
        min_distance
    }
}

fn create_photo_from_stdin() -> Photo {
    let lines = std::io::stdin().lines();
    let mut data = Vec::new();
    for line in  lines {
        match line {
            Ok(line) => {
                let mut v = Vec::new();
                for c in line.chars() {
                    match c {
                        '.' => v.push(Pixel::Air),
                        '#' => v.push(Pixel::Terran),
                        'X' => v.push(Pixel::Star),
                        _ => panic!(""),
                    }
                }
                data.push(v);
            }
            Err(_) => panic!(""),
        }
    }
    Photo {
        data: data,
    }
}
